bundles/org.openhab.binding.heos/src/main/java/org/openhab/binding/heos/internal/api/HeosAudioSink.java

```diff
@@ -77,42 +72,72 @@ public String getId() {
-                        handler.playURL(url + FileAudioStream.AAC_EXTENSION);
-                    } else {
-                        throw new UnsupportedAudioFormatException("HEOS only supports MP3, WAV and AAC.", audioFormat);
-                    }
-                } else {
-                    logger.warn("We do not have any callback url, so HEOS cannot play the audio stream!");
-                }
-            } else {
-                throw new UnsupportedAudioFormatException(
-                        "HEOS can only handle FixedLengthAudioStreams & URLAudioStream.", null);
+    protected void processAsynchronously(@Nullable AudioStream audioStream)
+            throws UnsupportedAudioFormatException, UnsupportedAudioStreamException {
+        if (!ThingHandlerHelper.isHandlerInitialized(handler)) {
+            logger.debug("HEOS speaker '{}' is not initialized - status is {}", handler.getThing().getUID(),
+                    handler.getThing().getStatus());
+            tryClose(audioStream);
+            return;
+        }
+
+        if (audioStream == null) {
+            return;
+        }
+
+        AudioFormat audioFormat = audioStream.getFormat();
+        if (!AudioFormat.MP3.isCompatible(audioFormat) && !AudioFormat.WAV.isCompatible(audioFormat)
+                && !AudioFormat.AAC.isCompatible(audioFormat)) {
+            tryClose(audioStream);
+            throw new UnsupportedAudioFormatException("HEOS speaker only supports MP3, WAV and AAC formats.",
+                    audioFormat);
+        }
+
+        String url;
+        if (audioStream instanceof URLAudioStream urlAudioStream) {
+            // it is an external URL, the speaker can access it itself and play it.
+            url = urlAudioStream.getURL();
+            tryClose(audioStream);
+        } else if (callbackUrl != null) {
+            StreamServed streamServed;
+            try {
+                streamServed = audioHTTPServer.serve(audioStream, 10, true);
+            } catch (IOException e) {
+                tryClose(audioStream);
+                throw new UnsupportedAudioStreamException(
+                        "HEOS was not able to handle the audio stream (cache on disk failed).", audioStream.getClass(),
+                        e);
+            }
+            url = callbackUrl + streamServed.url() + AudioStreamUtils.EXTENSION_SEPARATOR;
+            if (AudioFormat.MP3.isCompatible(audioFormat)) {
+                url += FileAudioStream.MP3_EXTENSION;
+            } else if (AudioFormat.WAV.isCompatible(audioFormat)) {
+                url += FileAudioStream.WAV_EXTENSION;
+            } else if (AudioFormat.AAC.isCompatible(audioFormat)) {
+                url += FileAudioStream.AAC_EXTENSION;
             }
+            streamServed.playEnd().thenRun(() -> this.playbackFinished(audioStream));
+        } else {
+            logger.warn("We do not have any callback url, so HEOS cannot play the audio stream!");
```

Comment:
Perhaps it could be slightly rephrased to avoid "we", something like: "No callback URL available, HEOS cannot play the audio stream." WDYT?

Reply:
No, it's not important. If you agree, you could change for not yet merged PR's, but I'm also okay to merge as is.

Reply:
Please merge and I will create 1 global PR for the 13 bindings to change this log when all currently opened PRs are merged.